File: daemon/rpytest_daemon/fixtures.py
"""Session fixture reuse management."""

import logging
import os
import subprocess
import sys
import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, Optional, Set

logger = logging.getLogger(__name__)


@dataclass
class FixtureState:
    """State of a session fixture."""
    name: str
    scope: str  # "session", "package", "module", "class", "function"
    created_at: float = 0.0
    last_used: float = 0.0
    use_count: int = 0
    teardown_pending: bool = False


@dataclass
class SessionState:
    """State of a pytest session with warm fixtures."""
    session_id: str
    repo_path: Path
    python_path: Path
    fixtures: Dict[str, FixtureState] = field(default_factory=dict)
    created_at: float = field(default_factory=time.time)
    last_run_at: float = 0.0
    total_runs: int = 0
    enabled: bool = False
    _lock: threading.Lock = field(default_factory=threading.Lock)

    def mark_fixture_used(self, name: str, scope: str = "session"):
        """Mark a fixture as used in the current run."""
        with self._lock:
            now = time.time()
            if name not in self.fixtures:
                self.fixtures[name] = FixtureState(
                    name=name,
                    scope=scope,
                    created_at=now,
                )
            self.fixtures[name].last_used = now
            self.fixtures[name].use_count += 1

    def mark_run_complete(self):
        """Mark a test run as complete."""
        with self._lock:
            self.last_run_at = time.time()
            self.total_runs += 1

    def get_stale_fixtures(self, max_age_seconds: float = 300) -> List[str]:
        """Get fixtures that haven't been used recently."""
        now = time.time()
        stale = []
        with self._lock:
            for name, state in self.fixtures.items():
                if now - state.last_used > max_age_seconds:
                    stale.append(name)
        return stale

    def to_dict(self) -> Dict:
        """Serialize to dict."""
        with self._lock:
            return {
                "session_id": self.session_id,
                "repo_path": str(self.repo_path),
                "created_at": self.created_at,
                "last_run_at": self.last_run_at,
                "total_runs": self.total_runs,
                "enabled": self.enabled,
                "fixtures": {
                    name: {
                        "name": f.name,
                        "scope": f.scope,
                        "created_at": f.created_at,
                        "last_used": f.last_used,
                        "use_count": f.use_count,
                    }
                    for name, f in self.fixtures.items()
                },
            }


class SessionFixtureManager:
    """Manages session fixture reuse across test runs.

    Safety guards:
    - Explicit opt-in required per context
    - Automatic teardown on file changes
    - Max age limit for fixtures
    - Isolation checks between tests
    """

    MAX_FIXTURE_AGE = 600  # 10 minutes default
    TEARDOWN_ON_CONFTEST_CHANGE = True

    def __init__(self):
        self._sessions: Dict[str, SessionState] = {}
        self._lock = threading.Lock()

    def create_session(
        self,
        context_id: str,
        repo_path: Path,
        python_path: Path,
    ) -> SessionState:
        """Create a new session state for a context."""
        with self._lock:
            session = SessionState(
                session_id=f"{context_id}-{int(time.time())}",
                repo_path=repo_path,
                python_path=python_path,
            )
            self._sessions[context_id] = session
            logger.info(f"Created session {session.session_id} for {context_id}")
            return session

    def get_session(self, context_id: str) -> Optional[SessionState]:
        """Get session state for a context."""
        with self._lock:
            return self._sessions.get(context_id)

    def enable_reuse(self, context_id: str) -> bool:
        """Enable fixture reuse for a context."""
        with self._lock:
            session = self._sessions.get(context_id)
            if not session:
                return False
            session.enabled = True
            logger.info(f"Enabled fixture reuse for {context_id}")
            return True

    def disable_reuse(self, context_id: str) -> bool:
        """Disable fixture reuse and teardown all fixtures."""
        with self._lock:
            session = self._sessions.get(context_id)
            if not session:
                return False
            session.enabled = False
            session.fixtures.clear()
            logger.info(f"Disabled fixture reuse for {context_id}")
            return True

    def is_reuse_enabled(self, context_id: str) -> bool:
        """Check if fixture reuse is enabled."""
        with self._lock:
            session = self._sessions.get(context_id)
            return session.enabled if session else False

    def invalidate_on_file_change(
        self,
        context_id: str,
        changed_files: List[Path],
    ) -> List[str]:
        """Invalidate fixtures based on file changes.

        Returns list of invalidated fixture names.
        """
        session = self.get_session(context_id)
        if not session or not session.enabled:
            return []

        invalidated = []

        # Check for conftest.py changes
        conftest_changed = any(
            f.name == "conftest.py" for f in changed_files
        )

        if conftest_changed and self.TEARDOWN_ON_CONFTEST_CHANGE:
            # All session fixtures need teardown
            with session._lock:
                invalidated = list(session.fixtures.keys())
                session.fixtures.clear()
            logger.info(
                f"Invalidated all fixtures for {context_id} due to conftest.py change"
            )
            return invalidated

        # Check for test file changes affecting specific fixtures
        # This would require deeper analysis of fixture dependencies
        # For now, invalidate all on any test file change
        test_files_changed = any(
            f.name.startswith("test_") or f.name.endswith("_test.py")
            for f in changed_files
        )

        if test_files_changed:
            # Keep session fixtures, but mark for potential re-evaluation
            logger.debug(
                f"Test files changed in {context_id}, fixtures retained"
            )

        return invalidated

    def cleanup_stale(
        self,
        context_id: str,
        max_age: float = None,
    ) -> List[str]:
        """Clean up stale fixtures.

        Returns list of cleaned up fixture names.
        """
        if max_age is None:
            max_age = self.MAX_FIXTURE_AGE

        session = self.get_session(context_id)
        if not session:
            return []

        stale = session.get_stale_fixtures(max_age)
        if stale:
            with session._lock:
                for name in stale:
                    session.fixtures.pop(name, None)
            logger.info(f"Cleaned up {len(stale)} stale fixtures for {context_id}")

        return stale

    def get_session_status(self, context_id: str) -> Optional[Dict]:
        """Get status of session fixture reuse."""
        session = self.get_session(context_id)
        if not session:
            return None
        return session.to_dict()

    def teardown_session(self, context_id: str):
        """Teardown all fixtures and remove session."""
        with self._lock:
            session = self._sessions.pop(context_id, None)
            if session:
                logger.info(f"Torn down session for {context_id}")

    def build_pytest_args(
        self,
        context_id: str,
        base_args: List[str],
    ) -> List[str]:
        """Build pytest arguments with session reuse settings.

        If reuse is enabled, adds arguments to preserve session state.
        """
        session = self.get_session(context_id)
        if not session or not session.enabled:
            return base_args

        # Add pytest arguments for session persistence
        # Note: This requires custom pytest plugin or conftest configuration
        args = base_args.copy()

        # The actual implementation would require a pytest plugin
        # that manages fixture caching across invocations
        # For now, we document the intended behavior

        return args


@dataclass
class FixtureConfig:
    """Configuration for fixture reuse behavior."""
    enabled: bool = False
    max_fixture_age_seconds: float = 600
    teardown_on_conftest_change: bool = True
    teardown_on_test_file_change: bool = False
    scopes_to_reuse: Set[str] = field(
        default_factory=lambda: {"session", "package"}
    )

    def to_dict(self) -> Dict:
        """Serialize to dict."""
        return {
            "enabled": self.enabled,
            "max_fixture_age_seconds": self.max_fixture_age_seconds,
            "teardown_on_conftest_change": self.teardown_on_conftest_change,
            "teardown_on_test_file_change": self.teardown_on_test_file_change,
            "scopes_to_reuse": list(self.scopes_to_reuse),
        }

    @classmethod
    def from_dict(cls, data: Dict) -> "FixtureConfig":
        """Deserialize from dict."""
        return cls(
            enabled=data.get("enabled", False),
            max_fixture_age_seconds=data.get("max_fixture_age_seconds", 600),
            teardown_on_conftest_change=data.get(
                "teardown_on_conftest_change", True
            ),
            teardown_on_test_file_change=data.get(
                "teardown_on_test_file_change", False
            ),
            scopes_to_reuse=set(data.get("scopes_to_reuse", ["session", "package"])),
        )

File: crates/rpytest-daemon/src/fixtures.rs
//! Session fixture reuse management.

use crate::models::{FixtureConfig, FixtureScope, FixtureState};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::SystemTime;
use tracing::debug;

/// Session state for a pytest session with warm fixtures.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionState {
    pub session_id: String,
    pub repo_path: String,
    pub python_path: String,
    pub fixtures: HashMap<String, FixtureState>,
    pub created_at: f64,
    pub last_run_at: f64,
    pub total_runs: u32,
    pub enabled: bool,
}

impl Default for SessionState {
    fn default() -> Self {
        let now = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        SessionState {
            session_id: String::new(),
            repo_path: String::new(),
            python_path: String::new(),
            fixtures: HashMap::new(),
            created_at: now,
            last_run_at: now,
            total_runs: 0,
            enabled: false,
        }
    }
}

/// Manages session fixture reuse across test runs.
#[derive(Debug, Clone)]
pub struct FixtureManager {
    /// Sessions by context_id
    sessions: Arc<Mutex<HashMap<String, SessionState>>>,
    /// Default configuration
    config: FixtureConfig,
    /// Maximum fixture age in seconds
    max_age_seconds: f64,
}

impl Default for FixtureManager {
    fn default() -> Self {
        FixtureManager {
            sessions: Arc::new(Mutex::new(HashMap::new())),
            config: FixtureConfig::default(),
            max_age_seconds: 600.0, // 10 minutes
        }
    }
}

impl FixtureManager {
    /// Create a new fixture manager.
    pub fn new() -> Self {
        Self::default()
    }

    /// Create a new session state for a context.
    pub fn create_session(
        &self,
        context_id: &str,
        repo_path: &PathBuf,
        python_path: &PathBuf,
    ) -> SessionState {
        let now = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        let session = SessionState {
            session_id: format!("{}-{}", context_id, now as u64),
            repo_path: repo_path.to_string_lossy().to_string(),
            python_path: python_path.to_string_lossy().to_string(),
            fixtures: HashMap::new(),
            created_at: now,
            last_run_at: now,
            total_runs: 0,
            enabled: false,
        };

        let mut sessions = self.sessions.lock().unwrap();
        sessions.insert(context_id.to_string(), session.clone());

        debug!("Created fixture session for {}", context_id);
        session
    }

    /// Get session state for a context.
    pub fn get_session(&self, context_id: &str) -> Option<SessionState> {
        let sessions = self.sessions.lock().unwrap();
        sessions.get(context_id).cloned()
    }

    /// Enable fixture reuse for a context.
    pub fn enable_reuse(&self, context_id: &str) -> bool {
        let mut sessions = self.sessions.lock().unwrap();
        if let Some(session) = sessions.get_mut(context_id) {
            session.enabled = true;
            debug!("Enabled fixture reuse for {}", context_id);
            true
        } else {
            false
        }
    }

    /// Disable fixture reuse and teardown all fixtures.
    pub fn disable_reuse(&self, context_id: &str) -> bool {
        let mut sessions = self.sessions.lock().unwrap();
        if let Some(session) = sessions.get_mut(context_id) {
            session.enabled = false;
            session.fixtures.clear();
            debug!("Disabled fixture reuse for {}", context_id);
            true
        } else {
            false
        }
    }

    /// Mark a fixture as used in the current run.
    pub fn mark_fixture_used(&self, context_id: &str, name: &str, scope: FixtureScope) {
        let now = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        let mut sessions = self.sessions.lock().unwrap();
        if let Some(session) = sessions.get_mut(context_id) {
            if !session.enabled {
                return;
            }

            let fixture =
                session
                    .fixtures
                    .entry(name.to_string())
                    .or_insert_with(|| FixtureState {
                        name: name.to_string(),
                        scope,
                        created_at: now,
                        last_used: now,
                        use_count: 0,
                        teardown_pending: false,
                    });
            fixture.last_used = now;
            fixture.use_count += 1;
        }
    }

    /// Mark a test run as complete.
    pub fn mark_run_complete(&self, context_id: &str) {
        let now = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        let mut sessions = self.sessions.lock().unwrap();
        if let Some(session) = sessions.get_mut(context_id) {
            session.last_run_at = now;
            session.total_runs += 1;
        }
    }

    /// Get fixtures that haven't been used recently.
    pub fn get_stale_fixtures(&self, context_id: &str) -> Vec<String> {
        let now = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        let mut stale = Vec::new();

        let mut sessions = self.sessions.lock().unwrap();
        if let Some(session) = sessions.get(context_id) {
            for (name, state) in &session.fixtures {
                if now - state.last_used > self.max_age_seconds {
                    stale.push(name.clone());
                }
            }
        }

        stale
    }

    /// Clear all sessions.
    pub fn clear_all(&self) {
        let mut sessions = self.sessions.lock().unwrap();
        sessions.clear();
    }

    /// Remove a specific context's session.
    pub fn remove_session(&self, context_id: &str) {
        let mut sessions = self.sessions.lock().unwrap();
        sessions.remove(context_id);
    }

    /// Get session count.
    pub fn session_count(&self) -> usize {
        self.sessions.lock().unwrap().len()
    }

    /// Configure fixture reuse settings.
    pub fn configure(&mut self, config: FixtureConfig) {
        self.config = config.clone();
        self.max_age_seconds = config.max_age_seconds;
    }

    /// Get current configuration.
    pub fn get_config(&self) -> &FixtureConfig {
        &self.config
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn test_create_session() {
        let manager = FixtureManager::new();
        let repo_path = PathBuf::from("/test/repo");
        let python_path = PathBuf::from("/usr/bin/python");

        let session = manager.create_session("ctx-1", &repo_path, &python_path);

        assert!(!session.session_id.is_empty());
        assert_eq!(session.repo_path, "/test/repo");
        assert!(manager.session_count() == 1);
    }

    #[test]
    fn test_enable_disable() {
        let manager = FixtureManager::new();
        let repo_path = PathBuf::from("/test/repo");
        let python_path = PathBuf::from("/usr/bin/python");

        manager.create_session("ctx-1", &repo_path, &python_path);
        assert!(manager.enable_reuse("ctx-1"));
        assert!(manager.disable_reuse("ctx-1"));
    }

    #[test]
    fn test_mark_fixture_used() {
        let manager = FixtureManager::new();
        let repo_path = PathBuf::from("/test/repo");
        let python_path = PathBuf::from("/usr/bin/python");

        manager.create_session("ctx-1", &repo_path, &python_path);
        manager.enable_reuse("ctx-1");
        manager.mark_fixture_used("ctx-1", "db_connection", FixtureScope::Session);

        let session = manager.get_session("ctx-1").unwrap();
        assert!(session.fixtures.contains_key("db_connection"));
        assert_eq!(session.fixtures["db_connection"].use_count, 1);
    }
}
